query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala

```diff
@@ -412,21 +407,23 @@ object RangeFunction {
 }
 
 object LastSampleFunction extends RangeFunction {
-  override def needsLastSample: Boolean = true
   def addedToWindow(row: TransientRow, window: Window): Unit = {}
   def removedFromWindow(row: TransientRow, window: Window): Unit = {}
   def apply(startTimestamp: Long,
             endTimestamp: Long,
             window: Window,
             sampleToEmit: TransientRow,
             queryConfig: QueryConfig): Unit = {
-    if (window.size > 1)
-      throw new IllegalStateException(s"Window had more than 1 sample. Possible out of order samples. Window: $window")
-    if (window.size == 0 || (endTimestamp - window.head.getLong(0)) > queryConfig.staleSampleAfterMs) {
-      sampleToEmit.setValues(endTimestamp, Double.NaN)
-    } else {
-      sampleToEmit.setValues(endTimestamp, window.head.getDouble(1))
+    for (i <- (window.size - 1) to 0 by -1) {
+      val row = window.apply(i)
+      val rowValue = row.getDouble(1)
+      //val rowTimestampMs = row.getLong(0)
+      if (!rowValue.isNaN ) {
+        sampleToEmit.setValues(endTimestamp, rowValue)
+        return
+      }
```

Comment:
Consider using cForRange or a while loop. `(window.size - 1) to 0 by -1` is much slower